diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,320 @@
+
+                    def _choose(tay_leaf, direct_leaf):
+                        assert tay_leaf.ndim == direct_leaf.ndim == cond.ndim, (
+                            f"tay_leaf.ndim: {tay_leaf.ndim},"
+                            f" direct_leaf.ndim: {direct_leaf.ndim},"
+                            f" cond.ndim: {cond.ndim}"
+                        )
+                        return jnp.where(cond, tay_leaf, direct_leaf)
+
+                    return jtu.tree_map(_choose, tay_out, direct_out)
+
+                # If all entries of h*gamma are below threshold, only compute tay_out
+                # otherwise, compute both tay_out and direct_out and select the
+                # correct one for each dimension
+                return lax.cond(
+                    eqxi.unvmap_all(cond), lambda _: tay_out, select_tay_or_direct, None
+                )
+
+        tree_with_coeffs = jtu.tree_map(recompute_coeffs_leaf, gamma, tay_coeffs)
+        outer = jtu.tree_structure(gamma)
+        inner = self._coeffs_structure
+        coeffs_with_tree = jtu.tree_transpose(outer, inner, tree_with_coeffs)
+        return coeffs_with_tree
+
+    def init(
+        self,
+        terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: LangevinTuple,
+        args: PyTree,
+    ) -> SolverState:
+        """Precompute _SolverState which carries the Taylor coefficients and the
+        SRK coefficients (which can be computed from h and the Taylor coefficients).
+        Some solvers of this type are FSAL, so _SolverState also carries the previous
+        evaluation of grad_f.
+        """
+        drift, diffusion = terms.terms
+        gamma, u, f = get_args_from_terms(terms)
+
+        h = drift.contr(t0, t1)
+        x0, v0 = y0
+
+        def _check_shapes(_c, _u, _x, _v):
+            assert _c.shape == _u.shape == _x.shape == _v.shape
+
+        assert jtu.tree_all(jtu.tree_map(_check_shapes, gamma, u, x0, v0))
+
+        tay_coeffs = jtu.tree_map(self._comp_taylor_coeffs_leaf, gamma)
+        # tay_coeffs have the same tree structure as gamma, with each leaf being a
+        # _Coeffs and the arrays have an extra trailing dimension of 6
+
+        coeffs = self._recompute_coeffs(h, gamma, tay_coeffs, -1.0)
+        rho = jtu.tree_map(lambda c, _u: jnp.sqrt(2 * c * _u), gamma, u)
+
+        state_out = SolverState(
+            h=h,
+            taylor_coeffs=tay_coeffs,
+            coeffs=coeffs,
+            rho=rho,
+            prev_f=f(x0),
+        )
+
+        return state_out
+
+    @staticmethod
+    @abc.abstractmethod
+    def _compute_step(
+        h: RealScalarLike,
+        levy,
+        x0: LangevinX,
+        v0: LangevinX,
+        langevin_args: _LangevinArgs,
+        coeffs: _Coeffs,
+        st: SolverState,
+    ) -> tuple[LangevinX, LangevinX, LangevinX, _ErrorEstimate]:
+        raise NotImplementedError
```

Comment:
Does this need to be a static method? Is there a reason we must demand that this method be parameterised only at the type level (i.e. via new subclasses) and not at the instance level (i.e. via attributes on `self`)?

Reply:
Fair point, I guess there's no reason not to allow that.